options(scipen=9999)
library(dplyr)

# codes to extract the genome-wide inbreeding prediction for each couple
# files read correspond to those from the moderately inbred simulation with a medium density array
# but the codes are valid for the other data sets

# comon variables or files to more than one method ####
ped <- read.table( "../predict/input/sim_ped" )
colnames(ped) <- c('id','ids','idd') 
rownames(ped) <- as.character(ped$id)
targ<- read.table( "../predict/input/sim_targets" )[,2]
targ<- ped[as.character(targ),]
rm(ped)
off <- targ
off_pairs <- paste(off$ids, off$idd, sep = "_")

# chromosomes and genome coverage
nchr=25
bim = read.table(file="../predict/input/sim_nontargets.bim")
cov = 0
for (i in 1:nchr){
  first = bim[bim$V1==i,][1,4]
  last  = bim[bim$V1==i,][nrow(bim[bim$V1==i,]),4]
  cov = cov + (last - first)
}
cov = cov/1000000
rm(bim, first, i, last, targ)

# extract information from outputs generated by each method & predict inbreeding    

# IBD_Haplo 9 classes (unphased) ####
coup<-read.table("../predict/ibd_haplo/output_unphased/1_unphased_2011.ids")[,c('V1','V4','V8')] #same for all chromosomes
colnames(coup)<- c('scoreset','V1','V2')
for (i in rownames(off)){
        ps = off[i,'ids']
        pd = off[i,'idd']
        score= which( (coup$V1 == ps & coup$V2 == pd) | (coup$V1 == pd & coup$V2 == ps) )
        off[i,'scoreset'] = score
}
off$IBD_Haplo9c = 0
nmrks = 0
for (i in 1:nchr){
  ffdir = paste0('../predict/ibd_haplo/output_unphased/',i,'_unphased_2011.qibd')
  ff<- read.table(ffdir, header = F)
  colnames(ff)<- c('scoreset','mrk','pos','S1','S2','S3','S4','S5','S6','S7','S8','S9')
  nmrks<- nmrks + max(ff$mrk)
  ff$ibdhaplo<- ff$S1 + 0.5*(ff$S3 + ff$S5 + ff$S7) + 0.25*ff$S8
  tmp0<- ff %>% group_by(scoreset) %>% summarize(ibdhaplo = sum(ibdhaplo, na.rm = TRUE))
  tmp0<- data.frame(tmp0)
  for(j in rownames(off)){
  off[j,'IBD_Haplo9c'] = off[j,'IBD_Haplo9c'] + tmp0[off[j,'scoreset'],'ibdhaplo']
  }
}
off$IBD_Haplo9c = off$IBD_Haplo9c/nmrks
off$scoreset = NULL
rm(coup, ff, ffdir, i, j, nmrks, pd, ps, score, tmp0)

# IBD_Haplo 15 classes (phased) ####
coup<-read.table("../predict/ibd_haplo/output_phased/1_phased_2011.ids")[,c('V1','V4','V8')] #same for all chromosomes
colnames(coup)<- c('scoreset','V1','V2')
for (i in rownames(off)){
        ps = off[i,'ids']
        pd = off[i,'idd']
        score= which( (coup$V1 == ps & coup$V2 == pd) | (coup$V1 == pd & coup$V2 == ps) )
        off[i,'scoreset'] = score
}
#15 states
#1     2     3     4     5     6     7     8     9     10    11    12    13    14    15
#1111, 1122, 1112, 1121, 1123, 1211, 1222, 1233, 1212, 1221, 1213, 1231, 1223, 1232, 1234

#9 condensed states
#1     2     3          4     5          6     7          8                    9
#1     2     3+4        5     6+7        8     9+10       11+12+13+14          15
#1111, 1122, 1112+1121, 1123, 1211+1222, 1233, 1212+1221, 1213+1231+1223+1232, 1234
off$IBD_Haplo15c = 0
nmrks = 0
for (i in 1:nchr){
  ffdir = paste0('../predict/ibd_haplo/output_phased/',i,'_phased_2011.qibd')
  ff<- read.table(ffdir, header = F)
  colnames(ff)<- c('scoreset','mrk','pos','S1','S2','S3','S4','S5','S6','S7','S8','S9','S10','S11','S12','S13','S14','S15')
  nmrks<- nmrks + max(ff$mrk)
  s1 = ff$S1
  s3 = ff$S3 + ff$S4
  s5 = ff$S6 + ff$S7
  s7 = ff$S9 + ff$S10
  s8 = ff$S11 + ff$S12 + ff$S13 + ff$S14
  ff$ibdhaplop <- s1 + 0.5*(s3 + s5 + s7) + 0.25*s8
  tmp0<- ff %>% group_by(scoreset) %>% summarize(ibdhaplop = sum(ibdhaplop, na.rm = TRUE))
  tmp0<- data.frame(tmp0)
  for(j in rownames(off)){
  off[j,'IBD_Haplo15c'] = off[j,'IBD_Haplo15c'] + tmp0[off[j,'scoreset'],'ibdhaplop']
  }
}
off$IBD_Haplo15c = off$IBD_Haplo15c/nmrks
off$scoreset=NULL
rm(coup, ff, ffdir, i, j, nmrks, pd, ps, score, tmp0, s1, s3, s5, s7, s8)

# GIBDLD ####
ffdir = "../predict/GIBDLD/prefix_genome.kinship.gz"
ff <- read.table(ffdir, header = F)
ff_pairs <- paste(ff$V2, ff$V4, sep = "_")
ff_pairs_reverse <- paste(ff$V4, ff$V2, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$GIBDLD[!is.na(match_idx_1)] <- ff$V15[ match_idx_1[ !is.na(match_idx_1) ]] 
off$GIBDLD[!is.na(match_idx_2)] <- ff$V15[ match_idx_2[ !is.na(match_idx_2) ]]
rm(ff, ff_pairs, ff_pairs_reverse, ffdir, match_idx_1, match_idx_2) 

# LocalNgsRelate ####
ffdir = "../predict/LocalNgsRelate/loc_locngsrelate.txt"
ff <- read.table(ffdir, header = F)
ff2 = aggregate(V5 ~ V1+V2 , data=ff, FUN=mean)
ff_pairs         <- paste(ff2$V1, ff2$V2, sep = "_")
ff_pairs_reverse <- paste(ff2$V2, ff2$V1, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$LocalNgsRelate[!is.na(match_idx_1)] <- ff2$V5[ match_idx_1[ !is.na(match_idx_1) ]]
off$LocalNgsRelate[!is.na(match_idx_2)] <- ff2$V5[ match_idx_2[ !is.na(match_idx_2) ]]
rm(ff, ff_pairs, ff_pairs_reverse, ff2, ffdir, match_idx_1, match_idx_2) 

# TRUFFLE ####
ffdir = "../predict/TRUFFLE/truffle.ibd"
ff <- read.table(ffdir, header = T)
ff$kin=0.25*ff$IBD1+0.5*ff$IBD2
ff_pairs         <- paste(ff$ID1, ff$ID2, sep = "_")
ff_pairs_reverse <- paste(ff$ID2, ff$ID1, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$TRUFFLE[!is.na(match_idx_1)] <- ff$kin[ match_idx_1[ !is.na(match_idx_1) ]]
off$TRUFFLE[!is.na(match_idx_2)] <- ff$kin[ match_idx_2[ !is.na(match_idx_2) ]]
rm(ff, ff_pairs, ff_pairs_reverse, ffdir, match_idx_1, match_idx_2)

# RZooRoH Mix7L ####
library(RZooRoH)
load('../predict/RZooRoH/ZR_Mix7L_haplocombi.RData')
# inbreeding predictions as the average of the inbreeding estimations over the 4 combinations of parental haplotypes
average<- (my.resu.pp@realized + my.resu.pm@realized + my.resu.mp@realized + my.resu.mm@realized)/4
# fields correspond to the value for each HBD class (HBD_R5, HBD_R25, HBD_R125, HBD_R625, HBD_R3126, HBD_R15625, nonHBD_R15625)
# we use the first classes (more recent ancestors, longer segments) to accummulate inbreeding 
ZooRoH_125 <- as.data.frame(rowSums(average[,1:3]))
ZooRoH_25  <- as.data.frame(rowSums(average[,1:2]))
ff_pairs <- my.data.pp@sample_ids
match_idx_1 <- match(off_pairs, ff_pairs)
off$`ZooRoH-125`[!is.na(match_idx_1)] <- ZooRoH_125[ match_idx_1[ !is.na(match_idx_1) ], 1]
off$`ZooRoH-25`[!is.na(match_idx_1)] <- ZooRoH_25[ match_idx_1[ !is.na(match_idx_1) ], 1]
rm(average, ff_pairs, match_idx_1,
   my.data.mm, my.data.mp, my.data.pm, my.data.pp,
   my.model.pp, my.resu.mm, my.resu.mp, my.resu.pm, my.resu.pp,
   ZooRoH_125, ZooRoH_25)

# RZooRoH 1R ####
#library(RZooRoH)
load('../predict/RZooRoH/ZR_1R_haplocombi.RData')
average<- (my.resu.pp@realized + my.resu.pm@realized + my.resu.mp@realized + my.resu.mm@realized)/4
# this is a single HBD-class model
ZooRoH_1R <- as.data.frame(average[,1])
ff_pairs <- my.data.pp@sample_ids
match_idx_1 <- match(off_pairs, ff_pairs)
off$`ZooRoH-1R`[!is.na(match_idx_1)] <- ZooRoH_1R[ match_idx_1[ !is.na(match_idx_1) ], 1]
rm(average, ff_pairs, match_idx_1,
   my.data.mm, my.data.mp, my.data.pm, my.data.pp,
   my.model.pp, my.resu.mm, my.resu.mp, my.resu.pm, my.resu.pp,
   ZooRoH_1R)
   
# PLINK ROH ####
ff.pp=read.table(file='../predict/ROH/fake_proh_pp.hom.indiv',header=T)
ff.pm=read.table(file='../predict/ROH/fake_proh_pm.hom.indiv',header=T)
ff.mp=read.table(file='../predict/ROH/fake_proh_mp.hom.indiv',header=T)
ff.mm=read.table(file='../predict/ROH/fake_proh_mm.hom.indiv',header=T)
# inbreeding predictions as the average of the inbreeding estimations over the 4 combinations of parental haplotypes
ff<- rbind(ff.pp, ff.pm, ff.mp, ff.mm) 
ff2 = aggregate(KB ~ IID , data=ff, FUN=sum)
roh = ff2$KB/(4*1000*cov)
ff_pairs <- ff2$IID
match_idx_1 <- match(off_pairs, ff_pairs)
off$`PLINK ROH` = 0
off$`PLINK ROH`[!is.na(match_idx_1)] <- roh[ match_idx_1[ !is.na(match_idx_1) ] ]
rm(ff, ff_pairs, ff.mm, ff.mp, ff.pm, ff.pp, ff2, match_idx_1, roh)
 
# phasedibd ####
ffdir = "../predict/phasedibd/loc_phasedibd.txt_clean"
ff <- read.table(ffdir, header = F)
ff$len = (ff$V2-ff$V1)/1000000                  # end - start position (bp) of each segment
ff2 = aggregate(len ~ V3+V4 , data=ff, FUN=sum) # accumulate segments for each couple
ff_pairs         <- paste(ff2$V3, ff2$V4, sep = "_")
ff_pairs_reverse <- paste(ff2$V4, ff2$V3, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$phasedibd = 0
off$phasedibd[!is.na(match_idx_1)] <- ff2$len[ match_idx_1[ !is.na(match_idx_1) ]] / (4*cov)
off$phasedibd[!is.na(match_idx_2)] <- ff2$len[ match_idx_2[ !is.na(match_idx_2) ]] / (4*cov)
rm(ff, ff_pairs, ff_pairs_reverse, ff2, ffdir, match_idx_1, match_idx_2)

# hap-IBD
ffdir = "../predict/hap-IBD/hapibd.ibd.gz"
ff <- read.table(ffdir, header = F)
ff2 = aggregate(V8 ~ V1+V3 , data=ff, FUN=sum)
ff_pairs         <- paste(ff2$V1, ff2$V3, sep = "_")
ff_pairs_reverse <- paste(ff2$V3, ff2$V1, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$`hap-IBD` = 0
off$`hap-IBD`[!is.na(match_idx_1)] <- ff2$V8[ match_idx_1[ !is.na(match_idx_1) ]] / (4*cov)
off$`hap-IBD`[!is.na(match_idx_2)] <- ff2$V8[ match_idx_2[ !is.na(match_idx_2) ]] / (4*cov)
rm(ff, ff_pairs, ff_pairs_reverse, ff2, ffdir, match_idx_1, match_idx_2)

# GERMLINE ####
ffdir = "../predict/GERMLINE/loc_germline_haplo.txt"
ff <- read.table(ffdir, header = F)
ff2 = aggregate(V13 ~ V2+V5 , data=ff, FUN=sum)
ff_pairs         <- paste(ff2$V2, ff2$V5, sep = "_")
ff_pairs_reverse <- paste(ff2$V5, ff2$V2, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$GERMLINE = 0
off$GERMLINE[!is.na(match_idx_1)] <- ff2$V13[ match_idx_1[ !is.na(match_idx_1) ]] / (4*cov)
off$GERMLINE[!is.na(match_idx_2)] <- ff2$V13[ match_idx_2[ !is.na(match_idx_2) ]] / (4*cov)
rm(ff, ff_pairs, ff_pairs_reverse, ff2, ffdir, match_idx_1, match_idx_2)

# Refined IBD
ffdir = "../predict/refinedibd/refinedibd.ibd.merged.gz"
ff <- read.table(ffdir, header = F)
ff2 = aggregate(V9 ~ V1+V3 , data=ff, FUN=sum)
ff_pairs         <- paste(ff2$V1, ff2$V3, sep = "_")
ff_pairs_reverse <- paste(ff2$V3, ff2$V1, sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$`Refined IBD` = 0
off$`Refined IBD`[!is.na(match_idx_1)] <- ff2$V9[ match_idx_1[ !is.na(match_idx_1) ]] / (4*cov)
off$`Refined IBD`[!is.na(match_idx_2)] <- ff2$V9[ match_idx_2[ !is.na(match_idx_2) ]] / (4*cov)
rm(ff, ff_pairs, ff_pairs_reverse, ff2, ffdir, match_idx_1, match_idx_2)

# UNI (as couple val / 2) ####
ffdir   = "../predict/ya_and_vr/ya.grm.gz"
ffdirid = "../predict/ya_and_vr/ya.grm.id"
ff      <-read.table(ffdir, header = F)
ffid    <-read.table(ffdirid, header = F)
ff_pairs <- paste(ffid$V2[ff$V1], ffid$V2[ff$V2], sep = "_")
ff_pairs_reverse <- paste(ffid$V2[ff$V2], ffid$V2[ff$V1], sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$UNI[!is.na(match_idx_1)] <- ff$V4[ match_idx_1[ !is.na(match_idx_1) ]] / 2
off$UNI[!is.na(match_idx_2)] <- ff$V4[ match_idx_2[ !is.na(match_idx_2) ]] / 2
rm(ff, ff_pairs, ff_pairs_reverse, ffdir, ffdirid, ffid, match_idx_1, match_idx_2)

# GRM (as couple val / 2) ####
ffdir   = paste0("../predict/ya_and_vr/vr.grm.gz")
ffdirid = paste0("../predict/ya_and_vr/vr.grm.id")
ff      <-read.table(ffdir, header = F)
ffid    <-read.table(ffdirid, header = F)
ff_pairs <- paste(ffid$V2[ff$V1], ffid$V2[ff$V2], sep = "_")
ff_pairs_reverse <- paste(ffid$V2[ff$V2], ffid$V2[ff$V1], sep = "_")
match_idx_1 <- match(off_pairs, ff_pairs_reverse)
match_idx_2 <- match(off_pairs, ff_pairs)
off$GRM[!is.na(match_idx_1)] <- ff$V4[ match_idx_1[ !is.na(match_idx_1) ]] / 2
off$GRM[!is.na(match_idx_2)] <- ff$V4[ match_idx_2[ !is.na(match_idx_2) ]] / 2
rm(ff, ff_pairs, ff_pairs_reverse, ffdir, ffdirid, ffid, match_idx_1, match_idx_2)

# PEDIGREE ####
library(pedigree)
ped<- read.table("../predict/input/out_pedigree.txt.gz")
colnames(ped)<- c('id','slim_ids','slim_idd')
# slim_ids 250-749   (t-14; founders of the recorded pedigree)
# slim_ids 750-1249  (t-13)
# slim_ids 1250-1749 (t-12)
# slim_ids 1750-2249 (t-11)
# slim_ids 2250-2749 (t-10)
# slim_ids 2750-2999 (t-9)
# slim_ids 3000-3249 (t-8)
# slim_ids 3250-3499 (t-7)
# slim_ids 3500-3749 (t-6)
# slim_ids 3750-3999 (t-5)
# slim_ids 4000-4249 (t-4)
# slim_ids 4250-4499 (t-3)
# slim_ids 4500-4749 (t-2)
# slim_ids 4750-4999 (parents; t-1)
# slim_ids 0-249     (offspring; t)) 0-99 are the target individuals of the study
pedigree = ped
pedigree$id<- paste0("tsk_",pedigree$id)
pedigree$slim_ids<- paste0("tsk_",pedigree$slim_ids)
pedigree$slim_idd<- paste0("tsk_",pedigree$slim_idd)
pedigree[pedigree$slim_ids == "tsk_-1",'slim_ids'] =0
pedigree[pedigree$slim_idd == "tsk_-1",'slim_idd'] =0
ord <- orderPed(pedigree)
pedigree <- pedigree[order(ord),]
pedF<- calcInbreeding(pedigree)
pedigree$Pedigree = pedF
pedigree[,c('slim_ids','slim_idd')] = NULL
pedigree<- pedigree[pedigree$id %in% paste0('tsk_',seq(0,99)),]
match_idx_1 <- match(off$id, pedigree$id)
off$Pedigree[match_idx_1] <- pedigree$Pedigree[ match_idx_1[ match_idx_1 ]] 

# Reference genome-wide inbreeding ####
load('../referenceF/glob_truef.RData')
  # if real data
  # load('../referenceF/real_glob_reference_HBDp.RData')
colnames(true_gwF)<- c('id','Recent Base Population',
		       'Intermediate Base Population','Distant Base Population')
data<- merge(off, true_gwF, by="id")
save(data, file = "global_sim_mediumdensity.RData")
